lib/src/core/guild/auto_moderation.dart

```diff
@@ -0,0 +1,173 @@
+import 'package:nyxx/src/core/snowflake.dart';
+
+abstract class IAutoModerationRule {
+  /// The id of this rule.
+  Snowflake get id;
+
+  /// The guild's id this rule is applied to.
+  Snowflake get guildId;
+
+  /// The name of this rule.
+  String get name;
+
+  /// The user which first created this rule.
+  Snowflake get creatorId;
+
+  /// The rule event type.
+  EventTypes get eventTypes;
+
+  /// The rule trigger type.
+  TriggerTypes get triggerTypes;
+
+  /// The trigger metadata.
+  ITriggerMetadata get triggerMetadata;
+
+  /// The actions which will execute when the rule is triggered.
+  List<IActionStructure> get actions;
+
+  /// Whether this rule is enabled.
+  bool get enabled;
+
+  /// The role ids that should not be affected by the rule (Maximum of 20).
+  List<Snowflake> get ignoredRoles;
+
+  /// The channel ids that should not be affected by the rule (Maximum of 50).
+  List<Snowflake> get ignoredChannels;
+}
+
+enum EventTypes {
+  messageSend(1);
+
+  final int value;
+  const EventTypes(this.value);
+
+  static EventTypes _fromValue(int value) {
+    switch (value) {
+      case 1:
+        return EventTypes.messageSend;
+      default:
+        throw Exception('Unknown enum value: $value');
+    }
+  }
+
+  @override
+  String toString() => 'EventTypes[$value]';
+}
+
+enum TriggerTypes {
+  /// Check if content contains words from a user defined list of keywords
+  keyword(1),
+
+  /// Check if content contains any harmful links
+  harmfulLink(2),
+
+  /// Check if content represents generic spam
+  spam(3),
+
+  /// Check if content contains words from internal pre-defined wordsets
+  keywordPreset(4);
+
+  final int value;
+  const TriggerTypes(this.value);
+
+  static TriggerTypes _fromValue(int value) {
+    switch (value) {
+      case 1:
+        return TriggerTypes.keyword;
+      case 2:
+        return TriggerTypes.harmfulLink;
+      case 3:
+        return TriggerTypes.spam;
+      case 4:
+        return TriggerTypes.keywordPreset;
+      default:
+        throw Exception('Unknown enum value: $value');
```

Comment:
Can we make new exception type like UnknownEnumValueException or something like that. And throw that here instead

Reply:
Yes, ofc

(Sorry for the delay, I was very busy last week)